src/esp/physics/RigidObject.cpp

```diff
@@ -213,5 +213,39 @@ Magnum::Matrix3 RigidObject::getInertiaMatrix() {
   return inertia;
 }
 
+//////////////////
+// VelocityControl
+Magnum::Matrix4 VelocityControl::integrateTransform(
+    const float dt,
+    const Magnum::Matrix4& objectTransform) {
+  // linear first
+  Magnum::Vector3 newTranslation = objectTransform.translation();
+  if (controllingLinVel) {
+    if (linVelIsLocal) {
+      newTranslation += objectTransform.rotation() *
+                        (linVel * dt);  // avoid local scaling of the velocity
+    } else {
+      newTranslation += linVel * dt;
+    }
+  }
+
+  Magnum::Matrix3 newRotationScaling = objectTransform.rotationScaling();
+  // then angular
+  if (controllingAngVel) {
+    Magnum::Quaternion q;
+    Magnum::Vector3 ha = angVel * dt * 0.5;  // vector of half angle
+    float l = ha.length();                   // magnitude
+    if (l > 0) {
+      float ss = std::sin(l) / l;
+      q = Magnum::Quaternion(
+          Magnum::Vector3{ha.x() * ss, ha.y() * ss, ha.z() * ss}, std::cos(l));
```

Comment:
```suggestion
          ha*ss, std::cos(l));
```

Same below.

Though ... to me this looks like an unfinished derivation, but can't quite point the finger on where exactly. Isn't it possible to get the quat out of it in a more direct way, without having to do the square root and trig operations on something that already was a scaled direction vector?

Reply:
Seems like this is just converting an angle-axis into a quaternion, and thus you can use: https://doc.magnum.graphics/magnum/classMagnum_1_1Math_1_1Quaternion.html#accba70d55affe884d8086110a130475a ?

Reply:
I wanted to suggest that at first, but even that didn't feel right... how's the angle actually represented in `angVel`? Is it the direction or the magnitude? (Sorry if my question is dumb :sweat_smile:)

Reply:
Yeah, angle-axis representation is another way to specify a log quaternion.  Looks like `angVel * dt` is a log quaternion, and this code is take the exponential of it to get the correspond unit quaternion (aka versor).  `Mn::Quaternion::rotation` also computes the exponential of a log quaternion, it just takes it in angle-axis format -- aka magnitude and direction format.

Reply:
Thanks for pointing this out. Much more concise. 😄